migration/migration.go

```diff
@@ -0,0 +1,179 @@
+package migration
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	logger "github.com/ipfs/go-log/v2"
+)
+
+var (
+	log = logger.Logger("migrations")
+
+	keyCurrentVersion = datastore.NewKey("version")
+)
+
+// Migrator ensures a datastore goes through all the needed
+// Migrations to upgrade its current version to the latest version.
+type Migrator struct {
+	ds         datastore.TxnDatastore
+	migrations map[int]Migration
+}
+
+// Migration runs a vA->v(A+1) migration.
+type Migration func(datastore.Txn) error
+
+// New returns a new Migrator.
+func New(ds datastore.TxnDatastore, migrations map[int]Migration) *Migrator {
+	m := &Migrator{
+		ds:         ds,
+		migrations: migrations,
+	}
+	return m
+}
+
+// Ensure detects the current datastore version, and runs all the known migrations
+// to upgrade to the latest known version.
+func (m *Migrator) Ensure() error {
+	currentVersion, emptyDS, err := m.getCurrentVersion()
+	if err != nil {
+		return fmt.Errorf("getting current version: %s", err)
+	}
+
+	targetVersion := m.getTargetVersion()
+
+	// If the database is empty, we can assume Powergate started fresh
+	// and requires no migration. Set current version to latest version
+	// in migrations script.
+	if emptyDS {
+		if err := m.bootstrapEmptyDatastore(targetVersion); err != nil {
+			return fmt.Errorf("bootstrapping empty database: %s", err)
+		}
+
+		return nil
+	}
+
+	log.Infof("Current datastore version is %d, target version %d", currentVersion, targetVersion)
+
+	if currentVersion == targetVersion {
+		return nil
+	}
+
+	if currentVersion > targetVersion {
+		return fmt.Errorf("migrations are forward only, current version %d, target version %d", currentVersion, targetVersion)
+	}
```

Comment:
Currently, this `Migrator` does forward only migrations. So check here for a situation in which the user might want to go back to an older tag of Powergate which isn't compatible with a newer datastore.

We can extend this lib to do forward and backward migrations. Now I did only forward since doesn't make sense for these big migrations of the PR to be rollbacked since that's impossible to do correctly.

We could define `Migrations` as rollback-able or not, and potentially allow backward migrations depending if the migration has a backward logic or not. That would be easy to do. Mentioning it just to consider for the future.